File: src/Python/28.StrStr.py

class Solution:
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        if len(needle) == "":
            return 0
        hl = len(haystack)
        nl = len(needle)
        i = 0
        j = 0
        while i < hl:
            if j == nl:
                return i - j
            if haystack[i] == needle[j]:
                i += 1
                j += 1
            else:
                i = i - j + 1
                j = 0
        if i == hl and j == nl:
            return i - j
        return -1

File: src/CSharp/1-100/28.StrStr.cs
namespace PeefyLeetCode.StrStr
{
    public class Solution
    {
        public int StrStr(string haystack, string needle)
        {
            if(string.IsNullOrEmpty(needle) == true)
                return 0;
            var n1 = haystack.Length;
            var n2 = needle.Length;
            var max = n1 - n2;
            var first = needle[0];
            for(var i = 0; i < n1 ;++i)
            {
                if (haystack[i] != first)
                {
                    while (++i <= max && haystack[i] != first) ;
                }
                if (i <= max)
                {
                    int j = i + 1;
                    int end = j + n2 - 1;
                    for (int k = 1; j < end && haystack[j]
                            == needle[k]; j++, k++) ;
                    if (j == end)
                    {
                        return i;
                    }
                }
            }
            return -1;
        }
    }
}